crates/cli-support/src/wit/outgoing.rs

```diff
@@ -337,6 +345,88 @@ impl InstructionBuilder<'_, '_> {
         Ok(())
     }
 
+    fn instruction_unwrap(&mut self) {
+        let decode_t = self
+            .instructions
+            .pop()
+            .expect("can't add Unwrap instruction as the first instruction");
+        self.input.push(AdapterType::I32);
+        self.input.push(AdapterType::I32);
+        self.instructions.push(InstructionData {
+            instr: Instruction::UnwrapResult {
+                table_and_alloc: None,
+            },
+            stack_change: StackChange::Modified {
+                popped: 2,
+                pushed: 0,
+            },
+        });
+        self.instructions.push(decode_t);
+    }
+
+    fn outgoing_result(&mut self, arg: &Descriptor) -> Result<(), Error> {
+        match arg {
+            Descriptor::Externref
+            | Descriptor::NamedExternref(_)
+            | Descriptor::I8
+            | Descriptor::U8
+            | Descriptor::I16
+            | Descriptor::U16
+            | Descriptor::I32
+            | Descriptor::U32
+            | Descriptor::F32
+            | Descriptor::F64
+            | Descriptor::I64
+            | Descriptor::U64
+            | Descriptor::Boolean
+            | Descriptor::Char
+            | Descriptor::Enum { .. }
+            | Descriptor::RustStruct(_)
+            | Descriptor::Ref(_)
+            | Descriptor::RefMut(_)
+            | Descriptor::CachedString
+            | Descriptor::String
+            | Descriptor::Option(_)
+            | Descriptor::Vector(_) => {
+                // These can be of varying lengths.
+                // The structure of ResultAbi is that the err is last, so if we generically
+                // push an instruction that will read the Ok type, for example
+                //
+                //     LoadRetptr [f64, u32, u32]
+                //     Decode { popped: 3 }
+                //
+                self._outgoing(arg)?;
+                // then push an instruction just before that one with two extra inputs
+                self.instruction_unwrap();
+                // then we end up with
+                //
+                //     LoadRetptr [f64, u32, u32, u32, u32]
+                //     UnwrapResult { popped: 2 } (the top two u32s)
+                //     Decode { popped: 3 } (f64, u32, u32)
+                //
+                // and the code for UnwrapResult is simple.
```

Comment:
This block here feels a bit brittle, can you add some more asserts here that everything is the shape as expected? For example this looks like it assumes that `self._outgoing(arg)` pushes 2 instructions unconditionally, so could this assert that 2 instructions are indeed pushed and the first one pushed is `LoadRetptr`?

Reply:
LoadRetptr should just be called the inputs list, it's prepended automatically later on. But you were right, this was assuming 1 instruction would be pushed, and String for example pushes 2 (one is a deferred free) so I managed to get Node to crash with SIGBUS freeing a null pointer because it popped the wrong values.

Reply:
Oop, called that one too early.

Reply:
Oh no, this is another thing. That bit is fixed.